app/models/order_item.rb

```diff
@@ -1,37 +1,70 @@
 class OrderItem < ActiveRecord::Base
-  # DB relationships
-  belongs_to :product
   belongs_to :order
+  belongs_to :product
+  has_one :seller, through: :product
+
+  before_save :product_has_stock?
 
-  # data validations
   validates :product_id, presence: true, numericality: { only_integer: true, greater_than: 0 }
   validates :order_id, presence: true, numericality: { only_integer: true, greater_than: 0 }
   validates :quantity_ordered, presence: true, numericality: { only_integer: true, greater_than: 0 }
-  # !R !W is there a way we can adjust this validation to check the product first?
+  validates :status, presence: true, inclusion: { in: %w(pending paid shipped canceled),
+    message: "%{value} is not a valid status" }
+  validate :order_item_is_unique?, on: [:create]
+
+  def more!
+    if product_has_stock? && product.stock > quantity_ordered
+      increment!(:quantity_ordered, 1)
+    else
+      errors.add(:product_stock, "Product must have available stock.")
+    end
+  end
+
+  def less!
+    if quantity_ordered > 1
+      update_column(:quantity_ordered, quantity_ordered - 1)
+    else
+      errors.add(:quantity_ordered, "You must remove this from your cart if you want to reduce its quantity any further.")
+    end
+  end
 
-  # scopes
-  scope :by_product, ->(prod_id) { where(product_id: prod_id) }
-  # this scope is unecessary, because product.order_items can be used.
-  # order_items_controller's #more will need to be rewritten to use the right syntax & to not reinvent the wheel.
+  def adjust_if_product_stock_changed!
+    max_quantity = product.stock
+    return if quantity_ordered <= max_quantity
+    update_column(:quantity_ordered, max_quantity)
+    errors[:product_stock] << "Quantity ordered was adjusted because not enough of this product was stock."
+  end
 
-  # this should probably be in helpers/order_items_helpers
-  # if this is only going to be used in the cart, maybe just use it inline inside the view
-  def remove_prompt_text
-    "Are you sure you want to remove this item (#{ quantity_ordered } #{ product.name }) from your cart?"
+  def remove_product_stock!
+    product.remove_stock!(quantity_ordered)
   end
 
-  # in the pull request:
-  # two sentences about if we're doing anything over the weekend
-  # and what our basic idea for monday is going to be
+  def mark_shipped
+    self.update(status: "shipped")
+  end
 
-  # maybe talk about buyers logging in w/ Jeremy?
+  def mark_canceled
+    self.update(status: "canceled")
 
-  def display_name
-    product.name
+    product.add_stock!(quantity_ordered)
   end
 
-  # note that we to talk about this more
-  def price # cost / item_cost
+  def total_item_price
     quantity_ordered * product.price
   end
+
+  def product_has_stock?
+    product.stock?
+  end
+
+  def quantity_too_high?
+    quantity_ordered >= product.stock
+  end
+
+  private
+    def order_item_is_unique?
```

Comment:
It doesn't seem like this method is actually a `?` method because it doesn't seem to be returning a T/F?